app/src/main/java/com/keylesspalace/tusky/fragment/NotificationsFragment.kt

```diff
@@ -0,0 +1,688 @@
+        layoutManager = LinearLayoutManager(context)
+        binding.recyclerView.layoutManager = layoutManager
+        binding.recyclerView.setAccessibilityDelegateCompat(
+            ListStatusAccessibilityDelegate(
+                binding.recyclerView,
+                this
+            ) { pos: Int ->
+                val notification = adapter.snapshot()[pos]
+                // We support replies only for now
+                if (notification is NotificationViewData) {
+                    notification.statusViewData
+                } else {
+                    null
+                }
+            }
+        )
+        binding.recyclerView.addItemDecoration(
+            DividerItemDecoration(
+                context,
+                DividerItemDecoration.VERTICAL
+            )
+        )
+
+        binding.recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
+            val actionButton = (activity as ActionButtonActivity).actionButton
+
+            override fun onScrolled(view: RecyclerView, dx: Int, dy: Int) {
+                actionButton?.let { fab ->
+                    if (!viewModel.uiState.value.showFabWhileScrolling) {
+                        if (dy > 0 && fab.isShown) {
+                            fab.hide() // Hide when scrolling down
+                        } else if (dy < 0 && !fab.isShown) {
+                            fab.show() // Show when scrolling up
+                        }
+                    } else if (!fab.isShown) {
+                        fab.show()
+                    }
+                }
+            }
+        })
+
+        binding.recyclerView.adapter = adapter.withLoadStateHeaderAndFooter(
+            header = NotificationsLoadStateAdapter { adapter.retry() },
+            footer = NotificationsLoadStateAdapter { adapter.retry() }
+        )
+
+        binding.buttonClear.setOnClickListener { confirmClearNotifications() }
+        binding.buttonFilter.setOnClickListener { showFilterMenu() }
+        (binding.recyclerView.itemAnimator as SimpleItemAnimator?)!!.supportsChangeAnimations =
+            false
+
+        /**
+         * Collect this flow to notify the adapter that the timestamps of the visible items have
+         * changed
+         */
+        val updateTimestampFlow = flow {
+            while (true) { delay(60000); emit(Unit) }
+        }.onEach {
+            layoutManager?.findFirstVisibleItemPosition()?.let { first ->
+                first == RecyclerView.NO_POSITION && return@let
+                val count = layoutManager!!.findLastVisibleItemPosition() - first
+                adapter.notifyItemRangeChanged(
+                    first,
+                    count,
+                    listOf(StatusBaseViewHolder.Key.KEY_CREATED)
+                )
+            }
+        }
+
+        lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
+                viewLifecycleOwner.lifecycleScope.launch {
```

Comment:
This is two distinct things.

First, the choice of the outermost scope. That should be `viewLifecycleOwner.lifecycleScope`, I've fixed that.

Second, why the nesting?

`.collect()` and similar are `suspend fun` and never return. So if you do this:

```
someScope.launch {
    someFlow.collect { /* do something */ }

    someOtherFlow.collect { /* do something else */ }
}
```

the `someOtherFlow.collect` line is unreachable because the `collect` before it never returns.

Collecting multiple flows in parallel needs to be done in separate coroutines (see also https://developer.android.com/topic/libraries/architecture/coroutines#lifecycle-aware).

Reply:
Yes I get that, but why is it not just

```
viewLifecycleOwner.lifecycleScope.launch {
    flow1.collect {}
}
viewLifecycleOwner.lifecycleScope.launch {
    flow2.collect {}
}
```

like e.g. [here](https://github.com/tuskyapp/Tusky/blob/develop/app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeActivity.kt#L407-L478)

Reply:
tl;dr: The existing code (e.g., in `ComposeActivity`) is unnecessarily consuming resources to get data for a view that the user cannot see, and should also be fixed.

Long version:

This:

```
viewLifecycleOwner.lifecycleScope.launch {
    flow1.collect {}
}
```

will continue collecting the flow until the activity or fragment that hosts it is destroyed. But a lot of the time they are not destroyed, they are paused, and then resumed. For example, when swiping between fragments in a viewpager, or launching an activity from an existing activity.

Imagine all the fragments in `MainActivity` have the code above to collect flows. Then the user clicks the compose button, and `ComposeActivity` is launched, `MainActivity` is paused (as are all the fragments it hosts).

But the coroutines launched by `MainActivity` and its hosted fragments are still running in the background, still collecting flows, still draining power.

I.e,. `viewLifecycleOwner` has nothing to do with the *visibility* of the view on screen.

By using `repeatOnLifecycle` and writing:

```
viewLifecycleOwner.lifecycleScope.launch {
    repeatOnLifecycle(Lifecycle.State.STARTED) {
        flow1.collect {}
    }
}
```

the coroutine that collects the flow will be cancelled when the activity or fragment is paused, and recreated when that activity or fragment resumes.

So if the activity or fragment is paused for whatever reason those coroutines won't be unncessarily consuming resources.

---

It's a bit of a stretch, but I occasionally see complaints from users that Tusky is draining their battery at an alarming rate. One of my hypotheses (that is currently *extremely* hard to test) is that this is a combination of flows being collected when they shouldn't be, and fragments being leaked.

Reply:
Ok, now I got you. Thanks for explaining 🙏

Reply:
Yes, that's why you use `STARTED`.

It's confusingly named, because `STARTED` runs from after `onStart()` to before `onPause()` (`CREATED` is the one that runs until `onStop()`). See https://developer.android.com/reference/androidx/lifecycle/Lifecycle.State which says:

> Started state for a LifecycleOwner. For an [android.app.Activity](https://developer.android.com/reference/android/app/Activity.html), this state is reached in two cases:

> - after [onStart](https://developer.android.com/reference/android/app/Activity.html#onStart()) call;

> - right before [onPause](https://developer.android.com/reference/android/app/Activity.html#onPause()) call.

That's a bit confusingly worded as well, as it could be read as implying that going:

```
onStart() [coroutine starts]--> onResume() --> [coroutine stops] --> onPause() --> [?] onResume()
```

does not recreate the coroutines on the `onPause() --> [?] onResume()` transition. But it does, so everything works as expected.